packages/story-editor/src/components/form/hierarchical/utils/index.js

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Creates an object composed of keys generated from the
+ * values of the `key` in each object.
+ *
+ * @param {Array.<Object>} arr The array to iterate over
+ * @param {string|number|undefined|null|boolean} key The key used to group the array items.
+ * @return {Object} The composed object
+ */
+const groupBy = (arr, key) =>
+  arr.reduce((prev, arrItem) => {
+    if (prev[arrItem[key]]) {
+      prev[arrItem[key]].push(arrItem);
+    } else {
+      prev[arrItem[key]] = [arrItem];
+    }
+    return prev;
+  }, {});
+
+/**
+ * Composes a tree from an array of options that are related.
+ * Can be parents or children as defined by the `parent` key.
+ *
+ * @param {Object} groupedOptionsByParent A object of options grouped by parent key.
+ * @param {Array.<Object>} options A flat array of options.
+ * @return {Array.<Object>} A tree of options
+ */
+const fillTree = (groupedOptionsByParent, options = []) => {
+  return options.map((option) => {
+    const children = groupedOptionsByParent[option.id];
+
+    return {
+      ...option,
+      options:
+        children && children.length
+          ? fillTree(groupedOptionsByParent, children)
+          : [],
+    };
+  });
+};
+
+/**
+ * Returns options in a tree form.
+ *
+ * Works similarly to Gutenberg Editor:
+ * https://github.com/WordPress/gutenberg/blob/3f9968e2815cfb56684c1acc9a2700d8e4a02726/packages/editor/src/utils/terms.js#L13-L40
+ *
+ * @param {Array} flatOptions Array of terms in flat format.
+ * @return {Array} Array of terms in tree format.
+ */
+export const buildOptionsTree = (flatOptions) => {
+  const formattedOptions = flatOptions.map((option) => ({
+    options: [],
+    parent: NO_PARENT_VALUE,
+    ...option,
+  }));
+
+  const groupedOptionsByParent = groupBy(formattedOptions, 'parent');
+
+  return fillTree(
+    groupedOptionsByParent,
+    groupedOptionsByParent[NO_PARENT_VALUE]
+  );
+};
+
+/**
+ * Filters an option and its children based on the label text. Only returns an option iff:
+ * 1. The option's label includes the `labelText`
+ * 2. The label of a child of the option includes the `labelText`
+ *
+ * Returns a list with a new option. This option's children (and all nested children) will
+ * also match the label text. If there are no children that match the label text then
+ * no children will be returned.
+ *
+ * @param {Object} option The option to check
+ * @param {string} labelText The text to match
+ * @return {Array.<Object>} A filtered list of options
+ */
+export const filterOption = (option, labelText = '') => {
+  const { options, ...rest } = option;
+  let newChildren = [];
+
+  // Find if any children match
+  if (options) {
+    for (let i = 0; i < options.length; i++) {
+      newChildren = newChildren.concat(filterOption(options[i], labelText));
+    }
+  }
+
+  // Find if current option matches
+  const matchesCurrentOption = option.label
+    .toLowerCase()
+    .includes(labelText?.trim().toLowerCase());
```

Comment:
Similar question here, do we expect `labelText` to ever be `null` ? Seems a little weird to use optional chaining when we default to a string.

Reply:
good call!